### PR TITLE
RPCDaemon: don't open uncomplete files at startup

### PR DESCRIPTION
**Pull request by AskAlexSharov**
Probably right way is: check `StageSnapshots > 0` as "all downloads done". Because:
- --prune.mode=minimal has not much files, but `preverified.toml` stays the  same
- downloader has logic to of prohibit_new_downloads.lock - downloading consider doen't even if `preverified.toml` has more files
- small files may get merged - while `preverified.toml` stays the same
- developer may generate new version of file (hash or len changed) - while `preverified.toml` stays the same
- Downloader's `Complete` flag can't be used - because e3 does download headers and bodies first, then Complete=true happens, and then we download transactions (Complete=false happens).

### CODE REVIEW

**Review by mh0lt — APPROVED** [review]

File: cmd/rpcdaemon/cli/config.go [hunk 1]
```diff
@@ -404,13 +407,30 @@ func RemoteServices(ctx context.Context, cfg *httpcfg.HttpCfg, logger log.Logger
...
 		// Configure sapshots
 		allSnapshots = freezeblocks.NewRoSnapshots(cfg.Snap, cfg.Dirs.Snap, 0, logger)
 		allBorSnapshots = freezeblocks.NewBorRoSnapshots(cfg.Snap, cfg.Dirs.Snap, 0, logger)
 		// To povide good UX - immediatly can read snapshots after RPCDaemon start, even if Erigon is down
 		// Erigon does store list of snapshots in db: means RPCDaemon can read this list now, but read by `remoteKvClient.Snapshots` after establish grpc connection
-		allSnapshots.OptimisticalyReopenFolder()
-		allBorSnapshots.OptimisticalyReopenFolder()
+		if doOptimisticOpen {
+			allSnapshots.OptimisticalyReopenFolder()
+			allBorSnapshots.OptimisticalyReopenFolder()
```
Comment (stevemilk) [inline cmd/rpcdaemon/cli/config.go:432]: the naming is OptimisticallReopenFolder, but the `ReopenFolder`  function it called doesn't open file in optimistic way.

**Review by mh0lt — APPROVED** [review]

File: erigon-lib/downloader/downloader.go [hunk 2]
```diff
@@ -2319,6 +2293,10 @@ func (d *Downloader) ReCalcStats(interval time.Duration) {
 
 	d.lock.Unlock()
 
+	if stats.Completed {
+		d.saveAllCompleteFlag()
```
Comment (stevemilk) [inline erigon-lib/downloader/downloader.go:2297]: There're two steps of downloading: 1. header chain segments(headers and bodies) 2. other segments(transactions...)
stats.Completed is true after the first step completed, but will be re-written to false once step2 starts.

should we wait all types of segments complete and then set the flag?

Comment (stevemilk) [reply erigon-lib/downloader/downloader.go:2297]: add some logs locally and here's logs test on sepolia:
```
Stats.Completed false
[INFO] [10-17|23:03:55.301] [1/6 OtterSync] Downloading              progress="(228/228 files) 99.71% - 2.7GB/2.7GB" rate=5.9MB/s time-left=0hrs:0m total-time=1m40s download-rate=5.9MB/s completion-rate=5.9MB/s alloc=552.7MB sys=653.8MB
Stats.Completed true
header and body snapshots downloaded completely
other snapshots downloading started
[INFO] [10-17|23:03:55.810] [snapshots] no metadata yet              files=90 list=domain/v1-accounts.0-128.bt,domain/v1-receipt.200-202.kv,v1-006710-006720-transactions.idx,v1-006730-006740-transactions.seg,domain/v1-accounts.192-200.bt,...
Stats.Completed false
```

Comment (AskAlexSharov) [reply erigon-lib/downloader/downloader.go:2297]: You right. Maybe i must use “StageSnapshots progress > 0”

**Review by stevemilk — APPROVED** [review]

**Comment by AskAlexSharov** [comment]
step2: https://github.com/erigontech/erigon/pull/12362

**Pull request merged**